horovod/run/elastic/registration.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2020 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import logging
+import threading
+
+from collections import defaultdict
+
+READY = 'READY'
+SUCCESS = 'SUCCESS'
+FAILURE = 'FAILURE'
+
+
+class WorkerStateRegistry(object):
+    def __init__(self, driver, host_manager, verbose=False):
+        self._driver = driver
+        self._host_manager = host_manager
+        self._lock = threading.Lock()
+        self._states = {}
+        self._workers = defaultdict(set)
+        self._barrier = None
+        self._rendezvous_id = 0
+        self._verbose = verbose
+        self._size = 0
+
+    def get_recorded_slots(self):
+        return self._states.keys()
+
+    def get(self, state):
+        return self._workers[state]
+
+    def count(self, state):
+        return len(self._workers[state])
+
+    def reset(self, size):
+        with self._lock:
+            logging.info('reset workers: {}'.format(size))
+            self._states.clear()
+            self._workers.clear()
+            self._barrier = threading.Barrier(parties=size, action=self._action)
+            self._rendezvous_id += 1
+            self._size = size
+
+    def size(self):
+        return self._size
+
+    def last_rendezvous(self):
+        return self._rendezvous_id
+
+    def record_ready(self, host, slot):
+        return self._record_state(host, slot, READY)
+
+    def record_success(self, host, slot):
+        return self._record_state(host, slot, SUCCESS)
+
+    def record_failure(self, host, slot):
+        return self._record_state(host, slot, FAILURE)
+
+    def _record_state(self, host, slot, state):
```

Comment:
I am running into the following situation:

- one worker fails
- the other worker resets
- registration registered all worker states so barrier fires (`_activate_hosts()`)
- then Spark kills the remaining worker because there were too many failures
- horovod has to timeout and does not shutdown immediately

```
[18:34:57        I registration.py:91 _record_state()] record state: special-bed85116180244240386e924d50a5a6e[0] = READY
[18:34:57        I rsh.py:70 rsh()] rsh exit code 1 for host special-f7aaa54d65dccc3b31a5088196cb3347 slot 0
[18:34:57        I registration.py:91 _record_state()] record state: special-f7aaa54d65dccc3b31a5088196cb3347[0] = FAILURE
[18:34:57        I registration.py:124 _on_workers_recorded()] all 2 workers recorded
[18:34:57        W discovery.py:107 blacklist()] blacklist failing host: special-f7aaa54d65dccc3b31a5088196cb3347
[18:34:57        I driver.py:154 _activate_hosts()] wait for available hosts: 2
20/05/05 18:35:01 WARN PythonRunner: Incomplete task 1.1 in stage 0 (TID 1) interrupted: Attempting to kill Python Worker
[18:35:01        I rsh.py:70 rsh()] rsh exit code None for host special-bed85116180244240386e924d50a5a6e slot 0
[18:35:01        I registration.py:89 _record_state()] key exists, reset barrier: special-bed85116180244240386e924d50a5a6e[0] = FAILURE
20/05/05 18:35:01 WARN TaskSetManager: Lost task 1.0 in stage 0.0 (TID 1, localhost, executor driver): TaskKilled (Stage cancelled)
```

Would you recommend to set the Spark failure limit to ∞ and rely on Horovods elastic failure count limit (is there one, I remember it being mentioned somewhere)? Or is this a situation the barrier reset should address?

I am happy to see this being addressed in a separate PR.

Reply:
For now, I think setting an unbounded failure limit is the right approach.  Added https://github.com/horovod/horovod/issues/1947 to track the reset limit in Horovod.